src/gar.lua

```diff
@@ -1577,4 +1577,271 @@ function gar.getPaginatedDelegations(address, cursor, limit, sortBy, sortOrder)
+--- @field delegateAddress string # The address of the delegate to redelegate stake from (required)
+--- @field sourceAddress string # The address of the gateway to redelegate stake from (required)
+--- @field targetAddress string # The address of the gateway to redelegate stake to (required)
+--- @field qty number # The amount of stake to redelegate - must be positive integer (required)
+--- @field currentTimestamp number # The current timestamp (required)
+--- @field vaultId string | nil # The vault id to redelegate from (optional)
+
+--- @class RedelegateStakeResult
+--- @field sourceGateway table # The updated gateway object that the stake was moved from
+--- @field targetGateway table # The updated gateway object that the stake was moved to
+--- @field redelegationFee number # The fee charged for the redelegation
+--- @field feeResetTimestamp number # The timestamp when the reldelegation fee will be reset
+--- @field redelegationsSinceFeeReset number # The number of redelegations the user has made since the last fee reset
+
+--- @param params RedelegateStakeParams
+--- @return RedelegateStakeResult
+function gar.redelegateStake(params)
+	local delegateAddress = params.delegateAddress
+	local targetAddress = params.targetAddress
+	local sourceAddress = params.sourceAddress
+	local stakeToTakeFromSource = params.qty
+	local currentTimestamp = params.currentTimestamp
+	local vaultId = params.vaultId
+
+	assert(type(stakeToTakeFromSource) == "number", "Quantity is required and must be a number")
+	assert(stakeToTakeFromSource > 0, "Quantity must be greater than 0")
+	assert(type(targetAddress) == "string", "Target address is required and must be a string")
+	assert(type(sourceAddress) == "string", "Source address is required and must be a string")
+	assert(type(delegateAddress) == "string", "Delegate address is required and must be a string")
+	assert(type(currentTimestamp) == "number", "Current timestamp is required and must be a number")
+	assert(sourceAddress ~= targetAddress, "Source and target gateway addresses must be different.")
+
+	local sourceGateway = gar.getGateway(sourceAddress)
+	local targetGateway = gar.getGateway(targetAddress)
+
+	assert(sourceGateway, "Source Gateway not found")
+	assert(targetGateway, "Target Gateway not found")
+	assert(
+		targetGateway.status ~= "leaving",
+		"Target Gateway is leaving the network and cannot have more stake delegated to it."
+	)
+	assert(targetGateway.settings.allowDelegatedStaking, "Target Gateway does not allow delegated staking.")
+	assert(
+		gar.delegateAllowedToStake(delegateAddress, targetGateway),
+		"This Gateway does not allow this delegate to stake."
+	)
+
+	local previousRedelegations = gar.getRedelegation(delegateAddress)
+
+	local redelegationFeeRate = gar.getRedelegationFee(delegateAddress).redelegationFeeRate
+
+	local redelegationFee = math.ceil(stakeToTakeFromSource * (redelegationFeeRate / 100))
+	local stakeToDelegate = stakeToTakeFromSource - redelegationFee
```

Comment:
So the redelegator chooses an amount of stake they are willing to participate in the move, and they basically get that amount minus any fees redelegated?

The other approach would be to move the amount they want to move and assess fees on TOP of that. Just want to make sure we're aligned with the intention from the white paper.

Reply:
Confirmed with Jon that this was the intended flow 👍